### PR TITLE
Library updates and cleanup

### PR DESCRIPTION
**Pull request by TacoTheDank**
### 1st commit
- Update buildTools 29.0.3 -> 30.0.3
~~- Use upstream subsamplingscaleimageview~~
- Update ExoPlayer 2.12.1 -> 2.12.2 ([changelog](https://github.com/google/ExoPlayer/blob/release-v2/RELEASENOTES.md))
~~- Update Flexmark 0.60.2 -> 0.62.2 ([changelog](https://github.com/vsch/flexmark-java/blob/master/VERSION.md))~~
- Update Kotlin 1.4.20 -> 1.4.21 ([changelog](https://github.com/JetBrains/kotlin/releases/tag/v1.4.21))
- Update Kotlin Coroutines 1.4.1 -> 1.4.2 ([changelog](https://github.com/Kotlin/kotlinx.coroutines/blob/master/CHANGES.md))
- Update Dagger 2.29.1 -> 2.30.1 ([changelog](https://github.com/google/dagger/releases))
- Update Joda-time 2.10.6 -> 2.10.8 ([changelog](https://www.joda.org/joda-time/changes-report.html))
- Update AndroidX Browser 1.2.0 -> 1.3.0 ([changelog](https://developer.android.com/jetpack/androidx/releases/browser))
- Update RxJava 2.2.19 -> 2.2.20 ([changelog](https://github.com/ReactiveX/RxJava/blob/2.x/CHANGES.md))
- Simplify MediaItem building

### 2nd commit
- Use some Compat methods and KTX extensions

### 3rd commit
- Miscellaneous cleanup

### CODE REVIEW

File: Kuroba/app/build.gradle [hunk 1]
```diff
@@ -269,17 +269,18 @@ dependencies {
...
 
-    implementation "com.google.android.exoplayer:exoplayer:${versions.exoplayer}"
+    implementation "com.google.android.exoplayer:exoplayer-core:${versions.exoplayer}"
+    implementation "com.google.android.exoplayer:exoplayer-ui:${versions.exoplayer}"
     implementation "com.squareup.okhttp3:okhttp:${versions.okhttp}"
     implementation "com.squareup.okhttp3:logging-interceptor:${versions.okhttp}"
     implementation "org.jsoup:jsoup:${versions.jsoup}"
     implementation 'pl.droidsonroids.gif:android-gif-drawable:1.2.21'
-    implementation 'com.davemorrissey.labs:subsampling-scale-image-view:3.10.0'
+    implementation 'com.github.davemorrissey:subsampling-scale-image-view:ebce8e0'
```
Comment (K1rakishou) [inline Kuroba/app/build.gradle]: What is the reason for this change? The latest version is 3.10.0 and "ebce8e0" is the latest commit https://github.com/davemorrissey/subsampling-scale-image-view/commit/ebce8e07ef57b163b68e198655587c6107b2ae67 and I don't really see why it should be used.

File: Kuroba/app/build.gradle [hunk 2]
```diff
@@ -269,17 +269,18 @@ dependencies {
...
     implementation 'org.greenrobot:eventbus:3.2.0'
     implementation 'org.nibor.autolink:autolink:0.10.0'
     implementation "com.google.code.gson:gson:${versions.gson}"
     // noinspection GradleDependency we want to get rid of this dependency eventually.
-    implementation 'com.vladsch.flexmark:flexmark:0.60.2'
+    implementation 'com.vladsch.flexmark:flexmark:0.62.2'
```
Comment (K1rakishou) [inline Kuroba/app/build.gradle]: I want to eventually get rid of this library and don't want to update it to the latest version since it will probably make the APK bigger. It's only used in like one or two places.

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/cell/PostCell.kt [hunk 3]
```diff
@@ -319,19 +319,15 @@ class PostCell : LinearLayout, PostCellInterface {
...
-    for (i in post!!.postImages.indices) {
-      if (!post!!.postImages[i].equalUrl(postImage)) {
-        continue
-      }
-
-      return if (isTextOnly) {
-        null
-      } else {
-        thumbnailViews[i]
+    return post!!.postImages.indices
```
Comment (K1rakishou) [inline Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/cell/PostCell.kt]: I don't think this change made the code more readable.

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/utils/MediaUtils.kt [hunk 4]
```diff
@@ -256,14 +259,10 @@ object MediaUtils {
...
-    var isJpegHeader = true
-    val size = Math.min(JPEG_HEADER.size, header.size)
+    val size = min(JPEG_HEADER.size, header.size)
 
-    for (i in 0 until size) {
-      if (header[i] != JPEG_HEADER[i]) {
-        isJpegHeader = false
-        break
-      }
+    val isJpegHeader = (0 until size).none {
```
Comment (K1rakishou) [inline Kuroba/app/src/main/java/com/github/k1rakishou/chan/utils/MediaUtils.kt]: I don't think this change made the code easier to read.

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/utils/MediaUtils.kt [hunk 5]
```diff
@@ -274,14 +273,10 @@ object MediaUtils {
...
-    val size = Math.min(PNG_HEADER.size, header.size)
+    val size = min(PNG_HEADER.size, header.size)
 
-    for (i in 0 until size) {
-      if (header[i] != PNG_HEADER[i]) {
-        isPngHeader = false
-        break
-      }
+    val isPngHeader = (0 until size).none {
+      header[it] != PNG_HEADER[it]
```
Comment (K1rakishou) [inline Kuroba/app/src/main/java/com/github/k1rakishou/chan/utils/MediaUtils.kt]: I don't think this change made the code easier to read.

File: Kuroba/core-model/src/main/java/com/github/k1rakishou/model/data/post/ChanPost.kt [hunk 6]
```diff
@@ -94,13 +94,9 @@ open class ChanPost(
 
   @Synchronized
   open fun allLoadersCompletedLoading(): Boolean {
-    for (loaderContentLoadState in onDemandContentLoadedMap.values) {
-      if (loaderContentLoadState.everythingLoaded()) {
-        return false
-      }
+    return onDemandContentLoadedMap.values.none {
+      it.everythingLoaded()
```
Comment (K1rakishou) [inline Kuroba/core-model/src/main/java/com/github/k1rakishou/model/data/post/ChanPost.kt]: I think there was a bug in the original code and it should be:

```kotlin
return onDemandContentLoadedMap.values.all { loaderContentLoadState ->
      loaderContentLoadState.everythingLoaded()
}
```

**Review by K1rakishou — CHANGES_REQUESTED** [review]
And please use merge instead of rebase.

File: Kuroba/app/build.gradle (hunk 1 above)
Comment (TacoTheDank) [reply Kuroba/app/build.gradle]: The master has two bugfixes that I feel are worth having: https://github.com/davemorrissey/subsampling-scale-image-view/compare/v3.10.0...master

File: Kuroba/app/build.gradle (hunk 2 above)
Comment (TacoTheDank) [reply Kuroba/app/build.gradle]: Fair enough 👍

File: Kuroba/core-model/src/main/java/com/github/k1rakishou/model/data/post/ChanPost.kt (hunk 6 above)
Comment (TacoTheDank) [reply Kuroba/core-model/src/main/java/com/github/k1rakishou/model/data/post/ChanPost.kt]: 👍

File: Kuroba/app/build.gradle (hunk 1 above)
Comment (K1rakishou) [reply Kuroba/app/build.gradle]: I think it should be fine with the current version (3.10.0). I haven't seen any of those crashes among the reported. We will update it once they release a stable version, no need to rush it.

Comment (TacoTheDank) [reply Kuroba/app/build.gradle]: Fair enough.

**Review by K1rakishou — APPROVED** [review]
